Monika After Story/game/special-effects.rpy

```diff
@@ -1,5 +1,670 @@
+            half_screen_height = screen_height / 2.0
+
+            # normalize the mouse position with the center of the screen
+            # 0.0 - left / bottom
+            # 1.0 - center
+            # 2.0 - right / top
+            mouse_x_factor = self.mouse_x / half_screen_width
+            mouse_y_factor = self.mouse_y / half_screen_height
+
+            container_offset_x = abs(self._container.offsets[0])
+            container_offset_y = abs(self._container.offsets[1])
+
+            # Our offsets consist of 4 parts:
+            # - base coords give offsets to x and y (depend on zoom level)
+            # - shift from the parallax effect (depends on mouse pos)
+            # - correction to the zoom effect (depends on mouse pos)
+            # - offsets from the container (depend on zoom level)
+            self._transform.xoffset = (
+                self._x*(1.0 + zoom_factor)
+                + available_x_shift*(1.0 - mouse_x_factor)
+                - zoom_factor*self.mouse_x
+                - container_offset_x*(1.0 + zoom_factor)
+            )
+            self._transform.yoffset = (
+                self._y*(1.0 + zoom_factor)
+                + available_y_shift*(1.0 - mouse_y_factor)
+                - zoom_factor*self.mouse_y
+                - container_offset_y*(1.0 + zoom_factor)
+            )
+
+            # Now update the displayable
+            self._transform.update()
+            renpy.redraw(self, 0.0)
+
+        def __getstate__(self):
+            """
+            This is used for pickling. Render objects cannot be pickled, so here we reset the value of _render to None.
+            The docs say CDD shouldn't keep Render, but we have to, as it works much better than
+            the built-in focus system which RenPy uses (for some reason it's more laggy and causes a few bugs).
+            Instead we keep the latest Render and check whether or not the pixel at the given x and y is opaque (apparently it's faster).
+
+            NOTE: The docs say CDD can be pickled, but it doesn't seem to be the case as Style objects inside them cannot be.
+                Nevertheless, I decided to handle this case so we can be sure that isn't our fault.
+            """
+            rv = super(ParallaxSprite, self).__getstate__()
+            rv["_render"] = None
+
+            return rv
+
+        def event(self, ev, x, y, st):
+            """
+            The event handler
+            TODO: allow to move around and zoom in/out with keyboard
+            """
+            if self._enable_events:
+                if ev.type == pygame.MOUSEMOTION:
+                    self.update_mouse_pos()
+                    self.update_offsets()
+
+                elif ev.type == pygame.MOUSEBUTTONDOWN:
+                    if self.min_zoom != self.max_zoom:
+                        if ev.button == 4:
+                            self.zoom += 0.1
+
+                        elif ev.button == 5:
```

Comment:
can you add comments denoting what kind of button press these are? I think these are just mouse wheel scrolls, right?

Reply:
Sure and yes, mouse wheel events.